src/server.rs

```diff
@@ -0,0 +1,294 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use tokio::net::UdpSocket;
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+
+use crate::ScuttleButt;
+
+/// Buffer size for UDP messages.
+/// This value is set to avoid UDP fragmentation.
+/// https://jvns.ca/blog/2017/02/07/mtu/
+const BUF_SIZE: usize = 1_472;
+
+/// UDP ScuttleButt server.
+pub struct ScuttleServer {
+    channel: UnboundedSender<ChannelMessage>,
+    scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    join_handle: JoinHandle<Result<(), anyhow::Error>>,
+}
+
+impl ScuttleServer {
+    /// Launch a new server.
+    ///
+    /// This will start the ScuttleButt server as a new Tokio background task.
+    pub fn spawn(address: impl Into<String>, seed_nodes: Vec<String>) -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let scuttlebutt = Arc::new(Mutex::new(ScuttleButt::with_node_id(address.into())));
+        let scuttlebutt_server = scuttlebutt.clone();
+
+        let join_handle = tokio::spawn(async move {
+            let mut server = UdpServer::new(rx, scuttlebutt_server).await?;
+            server.run().await
+        });
+
+        let scuttle_server = Self {
+            channel: tx,
+            scuttlebutt,
+            join_handle,
+        };
+
+        // Advertise the new node to all seed nodes.
+        for node_id in seed_nodes {
+            let _ = scuttle_server.gossip(node_id);
+        }
+
+        scuttle_server
+    }
+
+    /// Call a function with access to the [`ScuttleButt`].
+    pub async fn with_scuttlebutt<F, T>(&self, mut fun: F) -> T
+    where
+        F: FnMut(&ScuttleButt) -> T,
+    {
+        let scuttlebutt = self.scuttlebutt.lock().await;
+        fun(&scuttlebutt)
+    }
+
+    /// Call a function with mutable access to the [`ScuttleButt`].
+    ///
+    /// This will automatically gossip about the changes to all known nodes.
+    pub async fn with_scuttlebutt_mut<F, T>(&self, mut fun: F) -> T
+    where
+        F: FnMut(&mut ScuttleButt) -> T,
+    {
+        let mut scuttlebutt = self.scuttlebutt.lock().await;
+        let retval = fun(&mut scuttlebutt);
+
+        // Broadcast changes to all nodes after mutation.
+        for node in scuttlebutt.cluster_state_map.nodes() {
```

Comment:
Gossiping to all nodes would defy the purpose.

The point of that kind of gossip algorithm is to keep a bounded budget of network use.
We want a task that triggers a gossip round periodically (every second or so). 

Gossipping here means:
- picking 3 nodes randomly
- gossipping to them

You can check the cassandra links to see their refinements.

a) They pick from the alive node pool first.
b) They pick from a special seed + alive pool.
c) Sometimes they try to contact a dead node

We can pick nodes at random for the moment as we do not have a notion of liveness yet.
The algorithm will still be correct.

Reply:
we are within that map today. It might not be a good idea to send a gossip message to ourselves.

Reply:
> Gossiping to all nodes would defy the purpose.

That actually makes a lot of sense, yeah. I wasn't aware of the exact behavior but I think my latest commit should work much better.

In that case the whole `with_` stuff wouldn't really be necessary anymore. However this still "hides" the mutex lock. Do you prefer a `with_` method over exposing the `Arc<Mutex<_>>` publicly (or a `MutexGuard<_>` I suppose)?

I don't think I do, but you made it sound like you're quite opposed to having the lock interface public.

Reply:
I like the idea of passing a function.